[HLSL] Implement the `reflect` HLSL function

Icohedron opened the pull request:
Fixes #99152

Tasks completed:

- Implement `reflect` in `clang/lib/Headers/hlsl/hlsl_intrinsics.h`
- Implement the `reflect` SPIR-V target built-in in `clang/include/clang/Basic/BuiltinsSPIRV.td`
- Add a SPIR-V fast path in `clang/lib/Headers/hlsl/hlsl_detail.h` in the form
 ```c++
#if (__has_builtin(__builtin_spirv_reflect))
  return __builtin_spirv_reflect(...);
 #else
   return ...; // regular behavior
 #endif
 ```
- Add codegen for the SPIR-V `reflect` built-in to `EmitSPIRVBuiltinExpr` in `clang/lib/CodeGen/CGBuiltin.cpp`
- Add HLSL codegen tests to `clang/test/CodeGenHLSL/builtins/reflect.hlsl`
- Add SPIR-V built-in codegen tests to `clang/test/CodeGenSPIRV/Builtins/reflect.c`
- Add sema tests to `clang/test/SemaHLSL/BuiltIns/reflect-errors.hlsl`
- Add SPIR-V sema tests to `clang/test/CodeGenSPIRV/Builtins/reflect-errors.c`
- Create the `int_spv_reflect` intrinsic in `llvm/include/llvm/IR/IntrinsicsSPIRV.td`
- In `llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp` create the `reflect` lowering and map it to `int_spv_reflect` in `SPIRVInstructionSelector::selectIntrinsic`
- Create a SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll`

Additional tasks completed:

- Implement sema check for the `reflect` SPIR-V built-in in `clang/lib/Sema/SemaSPIRV.cpp`
  - Required for HLSL codegen to work via the SPIR-V fast path, because the types defined in `clang/include/clang/Basic/BuiltinsSPIRV.td` are being overridden
- Create SPIR-V backend error test case in `llvm/test/CodeGen/SPIRV/opencl/reflect-error.ll`
  - Since `reflect` is only available in the GLSL extended instruction set, using it in OpenCL should result in an error

Incomplete tasks:

- Create SPIR-V backend test case in `llvm/test/CodeGen/SPIRV/opencl/reflect.ll`
  - An OpenCL test is not applicable in this case because the [OpenCL SPIR-V extended instruction set](https://registry.khronos.org/SPIR-V/specs/unified1/OpenCL.ExtendedInstructionSet.100.html) does not include a `reflect` function

Icohedron committed:
Implement the `reflect` HLSL function

farzonl commented on clang/test/CodeGenHLSL/builtins/reflect.hlsl:
```diff
@@ -0,0 +1,195 @@
+// SPVCHECK-SAME: <2 x half> noundef nofpclass(nan inf) [[I:%.*]], <2 x half> noundef nofpclass(nan inf) [[N:%.*]]) local_unnamed_addr #[[ATTR0]] {
+// SPVCHECK-NEXT:  [[ENTRY:.*:]]
+// SPVCHECK-NEXT:    [[SPV_REFLECT_I:%.*]] = tail call reassoc nnan ninf nsz arcp afn <2 x half> @llvm.spv.reflect.v2f16(<2 x half> [[I]], <2 x half> [[N]])
+// SPVCHECK-NEXT:    [[SPLAT_SPLAT:%.*]] = shufflevector <2 x half> [[SPV_REFLECT_I]], <2 x half> poison, <2 x i32> zeroinitializer
```
Why are we returning a vector splat?

Icohedron replied:
I am unsure where the vector splat is coming from.

farzonl replied:
I don't see any vector splats in `clang/test/CodeGenSPIRV/Builtins/reflect.c` but I do in the CHECK and SPVCHECKs of this file so your issue seems specific to `hlsl_detail.h`

farzonl replied:
Oh and the SPLAT makes sense for the CHECK case because dot product is a scalar. It just doesn't make sense when calling the reflect SPIRV, clang builtin.

inbelic commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll:
```diff
@@ -0,0 +1,33 @@
+; RUN: llc -O0 -mtriple=spirv-unknown-unknown %s -o - | FileCheck %s
```
Out of curiousity, what happens when we use the triple `-mtriple=spirv32-unknown-unknown`? This should change the environment to be `OpenCL`, will it fail because there isn't a CL equivalent?

Icohedron replied:
Changing the triple to `-mtriple=spirv32-unknown-unknown` does indeed cause the test to fail at instruction selection

```
********************
FAIL: LLVM :: CodeGen/SPIRV/hlsl-intrinsics/reflect.ll (282 of 582)
******************** TEST 'LLVM :: CodeGen/SPIRV/hlsl-intrinsics/reflect.ll' FAILED ********************
Exit Code: 2

Command Output (stderr):
--
RUN: at line 1: /workspace/feature-reflect/build/bin/llc -O0 -mtriple=spirv32-unknown-unknown /workspace/feature-reflect/llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll -o - | /workspace/feature-reflect/build/bin/FileCheck /workspace/feature-reflect/llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll
+ /workspace/feature-reflect/build/bin/FileCheck /workspace/feature-reflect/llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll
+ /workspace/feature-reflect/build/bin/llc -O0 -mtriple=spirv32-unknown-unknown /workspace/feature-reflect/llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll -o -
LLVM ERROR: cannot select: %6:id(<2 x s64>) = G_INTRINSIC intrinsic(@llvm.spv.reflect), %0:vfid(<2 x s64>), %1:vfid(<2 x s64>) (in function: reflect_half4)
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /workspace/feature-reflect/build/bin/llc -O0 -mtriple=spirv32-unknown-unknown /workspace/feature-reflect/llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll -o -
1.      Running pass 'Function Pass Manager' on module '/workspace/feature-reflect/llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll'.
2.      Running pass 'InstructionSelect' on function '@reflect_half4'
 #0 0x00005a7a8f13318f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /workspace/feature-reflect/llvm/lib/Support/Unix/Signals.inc:804:3
 #1 0x00005a7a8f130a5f llvm::sys::RunSignalHandlers() /workspace/feature-reflect/llvm/lib/Support/Signals.cpp:105:20
 #2 0x00005a7a8f130db6 SignalHandler(int) /workspace/feature-reflect/llvm/lib/Support/Unix/Signals.inc:417:1
 #3 0x000071ee82242520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x000071ee822969fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x000071ee82242476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x000071ee822287f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00005a7a8f08b463 (/workspace/feature-reflect/build/bin/llc+0x3631463)
 #8 0x00005a7a8f664b32 reportGISelDiagnostic(llvm::DiagnosticSeverity, llvm::MachineFunction&, llvm::TargetPassConfig const&, llvm::MachineOptimizationRemarkEmitter&, llvm::MachineOptimizationRemarkMissed&) /workspace/feature-reflect/llvm/lib/CodeGen/GlobalISel/Utils.cpp:248:23
 #9 0x00005a7a8f666a55 llvm::DiagnosticInfoOptimizationBase::~DiagnosticInfoOptimizationBase() /workspace/feature-reflect/llvm/include/llvm/IR/DiagnosticInfo.h:486:7
#10 0x00005a7a8f666a55 llvm::DiagnosticInfoMIROptimization::~DiagnosticInfoMIROptimization() /workspace/feature-reflect/llvm/include/llvm/CodeGen/MachineOptimizationRemarkEmitter.h:31:7
#11 0x00005a7a8f666a55 llvm::MachineOptimizationRemarkMissed::~MachineOptimizationRemarkMissed() /workspace/feature-reflect/llvm/include/llvm/CodeGen/MachineOptimizationRemarkEmitter.h:85:7
#12 0x00005a7a8f666a55 llvm::reportGISelFailure(llvm::MachineFunction&, llvm::TargetPassConfig const&, llvm::MachineOptimizationRemarkEmitter&, char const*, llvm::StringRef, llvm::MachineInstr const&) /workspace/feature-reflect/llvm/lib/CodeGen/GlobalISel/Utils.cpp:277:1
#13 0x00005a7a8f5ce077 llvm::SmallVectorTemplateCommon<std::tuple<llvm::MachineBasicBlock*, llvm::MachineBasicBlock**, llvm::MachineBasicBlock**>, void>::begin() /workspace/feature-reflect/llvm/include/llvm/ADT/SmallVector.h:267:45
#14 0x00005a7a8f5ce077 llvm::SmallVectorTemplateCommon<std::tuple<llvm::MachineBasicBlock*, llvm::MachineBasicBlock**, llvm::MachineBasicBlock**>, void>::end() /workspace/feature-reflect/llvm/include/llvm/ADT/SmallVector.h:269:32
#15 0x00005a7a8f5ce077 llvm::SmallVector<std::tuple<llvm::MachineBasicBlock*, llvm::MachineBasicBlock**, llvm::MachineBasicBlock**>, 8u>::~SmallVector() /workspace/feature-reflect/llvm/include/llvm/ADT/SmallVector.h:1202:24
#16 0x00005a7a8f5ce077 llvm::po_iterator<llvm::MachineFunction*, llvm::SmallPtrSet<llvm::MachineBasicBlock*, 8u>, false, llvm::GraphTraits<llvm::MachineFunction*>>::~po_iterator() /workspace/feature-reflect/llvm/include/llvm/ADT/PostOrderIterator.h:97:7
#17 0x00005a7a8f5ce077 llvm::InstructionSelect::selectMachineFunction(llvm::MachineFunction&) /workspace/feature-reflect/llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp:225:5
#18 0x00005a7a8f5cf70c operator() /workspace/feature-reflect/llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp:145:59
#19 0x00005a7a8f5cf70c ~scope_exit /workspace/feature-reflect/llvm/include/llvm/ADT/ScopeExit.h:46:19
#20 0x00005a7a8f5cf70c llvm::InstructionSelect::runOnMachineFunction(llvm::MachineFunction&) /workspace/feature-reflect/llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp:157:1
#21 0x00005a7a8f5cf70c llvm::InstructionSelect::runOnMachineFunction(llvm::MachineFunction&) /workspace/feature-reflect/llvm/lib/CodeGen/GlobalISel/InstructionSelect.cpp:134:6
#22 0x00005a7a8dfbf72a llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) /workspace/feature-reflect/llvm/lib/CodeGen/MachineFunctionPass.cpp:94:33
#23 0x00005a7a8e576eca llvm::FPPassManager::runOnFunction(llvm::Function&) /workspace/feature-reflect/llvm/lib/IR/LegacyPassManager.cpp:1416:7
#24 0x00005a7a8e577119 llvm::ilist_detail::node_base_prevnext<llvm::ilist_node_base<true, void>, true>::getNext() const /workspace/feature-reflect/llvm/include/llvm/ADT/ilist_node_base.h:42:38
#25 0x00005a7a8e577119 llvm::ilist_node_impl<llvm::ilist_detail::node_options<llvm::Function, true, false, void, false, void>>::getNext() /workspace/feature-reflect/llvm/include/llvm/ADT/ilist_node.h:117:66
#26 0x00005a7a8e577119 llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Function, true, false, void, false, void>, false, false>::operator++() /workspace/feature-reflect/llvm/include/llvm/ADT/ilist_iterator.h:187:25
#27 0x00005a7a8e577119 llvm::FPPassManager::runOnModule(llvm::Module&) /workspace/feature-reflect/llvm/lib/IR/LegacyPassManager.cpp:1451:22
#28 0x00005a7a8e5779c0 runOnModule /workspace/feature-reflect/llvm/lib/IR/LegacyPassManager.cpp:1528:7
#29 0x00005a7a8e5779c0 llvm::legacy::PassManagerImpl::run(llvm::Module&) /workspace/feature-reflect/llvm/lib/IR/LegacyPassManager.cpp:539:55
#30 0x00005a7a8d4f741f compileModule(char**, llvm::LLVMContext&) /workspace/feature-reflect/llvm/tools/llc/llc.cpp:753:34
#31 0x00005a7a8d4f7de6 main /workspace/feature-reflect/llvm/tools/llc/llc.cpp:411:35
#32 0x000071ee82229d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#33 0x000071ee82229e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#34 0x00005a7a8d4e9e65 _start (/workspace/feature-reflect/build/bin/llc+0x1a8fe65)
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /workspace/feature-reflect/build/bin/FileCheck /workspace/feature-reflect/llvm/test/CodeGen/SPIRV/hlsl-intrinsics/reflect.ll
```

farzonl replied:
This won't be a problem for opencl because they don't have a reflect api to trigger this crash. This could be a problem if folks try to target spirv32/spirv64 via c because we just exposed a clang builtin. Since we just enabled this  and are so far the only users of the spirv target builtins i'm not to worried about this.

That said there are a few ways we could address this.
1. remove the reflect spirv target builtin and wait for this pr to merge: https://github.com/llvm/llvm-project/pull/122839/files then add an instcombine pattern for `I - 2 * N * dot(I, N);` to map to reflect.
2. ignore this case
3. lower the intrinsic back to `I - 2 * N * dot(I, N);` if its not targeting GLSL.  (This is my least favorite option).

farzonl commented on clang/test/CodeGenHLSL/builtins/reflect.hlsl:
```diff
@@ -0,0 +1,195 @@
+// CHECK-NEXT:    [[HLSL_DOT_I:%.*]] = tail call reassoc nnan ninf nsz arcp afn half @llvm.dx.fdot.v2f16(<2 x half> [[I]], <2 x half> [[N]])
+// CHECK-NEXT:    [[TMP0:%.*]] = extractelement <2 x half> [[I]], i64 0
+// CHECK-NEXT:    [[TMP1:%.*]] = extractelement <2 x half> [[N]], i64 0
+// CHECK-NEXT:    [[TMP2:%.*]] = fmul reassoc nnan ninf nsz arcp afn half [[TMP1]], 0xH4000
```
I'm a little confused here.  It looks like we are only multiplying 2 by the first element of `N`. I would have expected `2 *N` to be multiplied against both elements.

Icohedron replied:
Yea that's another issue I spotted and am troubleshooting right now

Icohedron replied:
It seems like all the math operations except for the dot product are being performed on only the first elements of each input vector.

farzonl replied:
I'm expecting the codegen to look like this
```llvmir
  %N.i0 = extractelement <2 x half> %N, i32 0
  %.i01 = fmul fast half 2.000000e+00, %N.i0
  %N.i1 = extractelement <2 x half> %N, i32 1
  %.i12 = fmul fast half 2.000000e+00, %N.i1
  %D    = half @llvm.dx.fdot.v2f16(<2 x half> %I, <2 x half> %N) 
  %.i03 = fmul fast half %.i01, %D
  %.i04 = fmul fast half %.i12, %D
 %Dv = shufflevector <2 x half> %i03 <2 x half> %i04
  %.i05 = fsub fast <2 x half> %I, %.Dv
  
```

inbelic commented on clang/lib/Sema/SemaSPIRV.cpp:
```diff
@@ -51,6 +51,38 @@ bool SemaSPIRV::CheckSPIRVBuiltinFunctionCall(unsigned BuiltinID,
+    }
+
+    QualType RetTy = ArgTyA;
+    TheCall->setType(RetTy);
```
FYI: this is the reason that it will fail to compile without having the diagnostics. Since the intrinsic types are being overloaded, you have to set it specifically here.

inbelic commented on clang/lib/Sema/SemaSPIRV.cpp:
```diff
@@ -51,6 +51,38 @@ bool SemaSPIRV::CheckSPIRVBuiltinFunctionCall(unsigned BuiltinID,
+
+    ExprResult A = TheCall->getArg(0);
+    QualType ArgTyA = A.get()->getType();
+    auto *VTyA = ArgTyA->getAs<VectorType>();
```
If you look through this file, there should be a helper function that will check what you are checking here

farzonl replied:
I think the helpers you are thinking about might be in SemaHLSL.cpp. SemaSPIRV.cpp is very barebones. If you think there is a helper that could be used across both files, we probably want to move it to a more generic place.

inbelic commented on clang/test/SemaHLSL/BuiltIns/reflect-errors.hlsl:
```diff
@@ -0,0 +1,33 @@
+
+float test_int_inputs(int p0, int p1) {
+  return reflect(p0, p1);
+  // expected-error@-1  {{call to 'reflect' is ambiguous}}
```
I am not sure I understand why this is the expected error message. Is it because it could cast to a float or half? That doesn't seem right

farzonl replied:
This is correct. see `ScalarOverloadResolution.hlsl` When there are more than one overload we get ambiguous. and when there is just one implementation, we get expected warning.

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2944,6 +2944,10 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
   case Intrinsic::spv_normalize:
     return selectExtInst(ResVReg, ResType, I, CL::normalize, GL::Normalize);
+  case Intrinsic::spv_reflect:
+    if (STI.isVulkanEnv()) // There is no CL equivalent of Reflect
```
I don't think this if statement does what you think it does.  As long as the intrinsic remains in the MIR you won't be able to proceed with codegen.

instead if you want to gracefully exit you should modify the GLSL version of  `selectExtInst ` to do something like this
```cpp
if (!STI.canUseExtInstSet(SPIRV::InstructionSet::GLSL_std_450)) {
    std::string DiagMsg;
    raw_string_ostream OS(DiagMsg);
    I.print(OS);
    DiagMsg = "Intrinsic selection not supported for this instruction set: " + DiagMsg;
    report_fatal_error(DiagMsg.c_str(), false);
}
//original behavior of selectExtInst
```

inbelic commented on clang/test/CodeGenSPIRV/Builtins/reflect.c:
```diff
@@ -0,0 +1,32 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
+
+// RUN: %clang_cc1 -O1 -triple spirv-pc-vulkan-compute %s -emit-llvm -o - | FileCheck %s
```
Do we also want to test a DirectX target? Or is the purpose of this test just to check we generate the intrinsic rather than expansion.

farzonl replied:
The clang codgen tests should be sufficient. Thats one of the advantages of moving the intrinsics into the header.

inbelic reviewed (approved):
This pr LGTM once the graceful exit comment is handled.

With respect to [this comment](https://github.com/llvm/llvm-project/pull/122992#discussion_r1917127000). Is the final solution to just gracefully exit or is this an intermediate step and an issue will be used to track an update using inst combine later?

Will a follow-up issue be created to move the useful `Sema` helpers into a common file?

farzonl commented:
> This pr LGTM once the graceful exit comment is handled.
> 
> With respect to [this comment](https://github.com/llvm/llvm-project/pull/122992#discussion_r1917127000). Is the final solution to just gracefully exit or is this an intermediate step and an issue will be used to track an update using inst combine later?

The graceful exit should be sufficient. We don't expose the clang builtin when targeting spirv32 or spirv64 and for now we are the only consumers of this intrinsic. Like I said in the last comment to only way to trigger this crash is with hand spun llvmir. To me making this not crash if you aren't in a supported environment is good enough.

farzonl commented:
> Will a follow-up issue be created to move the useful `Sema` helpers into a common file?

I'm fine if you want to do this as a follow up.

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3030,6 +3031,15 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
   case Intrinsic::spv_normalize:
     return selectExtInst(ResVReg, ResType, I, CL::normalize, GL::Normalize);
+  case Intrinsic::spv_reflect:
+    if (!STI.canUseExtInstSet(SPIRV::InstructionSet::InstructionSet::GLSL_std_450)) {
```
I was hoping to put this conditional in the GL overloaded version of selectExtInst so that other GL only extention instructions could benefit.

Icohedron replied:
Should I add a similar conditional to the OpenCL overloaded version of selectExtInst? Just to keep the OpenCL and GLSL functions similar.

farzonl replied:
Not a bad idea, but in another pr. so that the changes here are more focused to the reflect feature. Feel free to file an issue and identify some tests to add.

Icohedron replied:
I made an issue for this here #123847

farzonl commented on llvm/test/CodeGen/SPIRV/opencl/reflect-error.ll:
```diff
@@ -0,0 +1,22 @@
+; RUN: not %if spirv-tools %{ llc -O0 -mtriple=spirv64-unknown-unknown %s -o /dev/null 2>&1 -filetype=obj %}
+; RUN: not %if spirv-tools %{ llc -O0 -mtriple=spirv32-unknown-unknown %s -o /dev/null 2>&1 -filetype=obj %}
+
+; CHECK: LLVM ERROR: Intrinsic selection not supported for this instruction set: %{{.*}} = G_INTRINSIC intrinsic(@llvm.spv.reflect), %{{.*}}, %{{.*}}
```
This is my bad now that I'm seeing it  the lannguage is a bit obtuse and it doesn't look like instruction is encoded. We should simplify this error message.

Maybe:
```suggestion
; CHECK: LLVM ERROR: %{{.*}} = G_INTRINSIC intrinsic(@llvm.spv.reflect), % is only supported for GL instructions.
```
Or if you have something better go with that.

Icohedron replied:
I have changed the error message to read `LLVM ERROR: %{{.*}} = G_INTRINSIC intrinsic(@llvm.spv.reflect), %{{.*}}, %{{.*}} is only supported with the GLSL extended instruction set.`

farzonl commented on llvm/test/CodeGen/SPIRV/opencl/reflect-error.ll:
```diff
@@ -0,0 +1,22 @@
+
+define noundef <4 x float> @reflect_float4(<4 x float> noundef %a, <4 x float> noundef %b) {
+entry:
+  %spv.reflect = call <4 x float> @llvm.spv.reflect.f32(<4 x float> %a, <4 x float> %b)
```
This function will never be evaluated. The `report_fatal_error` means we stop codegen after line 10. Please delete.

Icohedron committed:
Add check for GLSL extended instruction set use

Certain SPIR-V instructions are only available in the GLSL extended
instruction set, such as "Reflect".

This commit adds a check to ensure the current SPIR-V target can use
the GLSL extended instruction set before selecting an instruction that
is only available from the GLSL extended instruction set.

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -34,6 +34,7 @@
 #include "llvm/IR/IntrinsicsSPIRV.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/raw_ostream.h"
```
Not sure why you need an include your changes to  `selectExtInst` are the same string stream as the default intrinsic error reporting case: https://github.com/llvm/llvm-project/blob/6518b121f037717fd211c36659f7b25266424719/llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp#L3124C5-L3128C48

Icohedron replied:
I don't recall adding that. I think my IDE did that for me :P 
I will remove it

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3030,6 +3039,8 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
   case Intrinsic::spv_normalize:
     return selectExtInst(ResVReg, ResType, I, CL::normalize, GL::Normalize);
+  case Intrinsic::spv_reflect:
+    return selectExtInst(ResVReg, ResType, I, GL::Reflect);
```
go to this line: https://github.com/llvm/llvm-project/blob/6518b121f037717fd211c36659f7b25266424719/llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp#L273

and remove the `[[maybe_unused]]`

farzonl reviewed (approved):
LGTM, but I want to see the `[[maybe_unused]]` and the `#include "llvm/Support/raw_ostream.h"` removed before you merge.

Icohedron committed:
Address PR comments

- Remove [[maybe_unused]] from selectExtInst
- Remove redundant include

Icohedron commented:
> Will a follow-up issue be created to move the useful Sema helpers into a common file?

I just created issue #123831 for this.

inbelic commented:
The build failures are because of this commit and the `reflect-errors.c` test. It seems that we wanted to delete this testcase as mentioned [here](https://github.com/llvm/llvm-project/pull/122992/files#r1920910435)? But since I am not sure on what the quick resolution is I will revert the change.
